doc/examples/denoise_ascm.py

```diff
@@ -0,0 +1,184 @@
+
+The adaptive soft coefficient matching (ASCM) as described in [Coupe11]_ is a
+improved extension of non-local means (NLMEANS) denoising. ASCM gives a better
+denoised images from two standard non-local means denoised versions of the
+original data with different degrees sharp feature preserved. Here, one
+denoised input is more "smooth" than the other (the easiest way to achieve this
+denoising is use ``non_local_means`` with two different patch radii).
+
+ASCM involves these basic steps
+
+* Computes wavelet decomposition of the noisy as well as denoised inputs
+
+* Combines the wavelets for the output image in a way that it takes it's
+  smoothness (low frequency components) from the input with larger smoothing,
+  and the sharp features (high frequency components) from the input with
+  less smoothing.
+
+This way ASCM gives us a well denoised output while preserving the sharpness
+of the image features, which is the primary goal in any denoising algorithm.
+
+Let us load the necessary modules
+"""
+
+import numpy as np
+import matplotlib.pyplot as plt
+import nibabel as nib
+from dipy.data import (fetch_sherbrooke_3shell,
+                       read_sherbrooke_3shell)
+from dipy.denoise.noise_estimate import estimate_sigma
+from time import time
+from dipy.denoise.non_local_means import non_local_means
+from dipy.denoise.adaptive_soft_matching import adaptive_soft_matching
+
+"""
+Choose one of the data from the datasets in DIPY
+"""
+
+fetch_sherbrooke_3shell()
+img, gtab = read_sherbrooke_3shell()
+
+data = img.get_data()
+affine = img.get_affine()
+
+mask = data[..., 0] > 80
+data = data[..., 1]
+
+print("vol size", data.shape)
+
+t = time()
+
+"""
+In order to generate the two pre-denoised versions of the data we will use the
+``non_local_means`` denoising. For ``non_local_means`` first we need to
+estimate the standard deviation of the noise. We use N=4 since the Sherbrooke
+dataset was acquired on a 1.5T Siemens scanner with a 4 array head coil.
+"""
+
+sigma = estimate_sigma(data, N=4)
+
+"""
+Non-local means with a smaller patch size which implies less smoothing, more
+sharpness
+"""
+
+den_small = non_local_means(
+    data,
+    sigma=sigma,
+    mask=mask,
+    patch_radius=1,
+    block_radius=1,
+    rician=True)
+
+"""
+Non-local means with larger patch size which implies more smoothing, less
+sharpness
+"""
+
+den_large = non_local_means(
+    data,
+    sigma=sigma,
+    mask=mask,
+    patch_radius=2,
+    block_radius=1,
+    rician=True)
+
+"""
+Now we perform the adaptive soft coefficient matching. Empirically we set the
+adaptive parameter in ascm to be the average of the local noise variance,
+in this case the sigma itself.
+"""
+
+den_final = np.array(adaptive_soft_matching(data, den_small, den_large,
+                                            sigma[0]))
+
+print("total time", time() - t)
+
+"""
+Plot the axial slice of the data, it's denoised output and the residual
```

Comment:
To access the quality of this denoising procedure, we plot the an axial slice of the original data, it's denoised output and residuals.